interfaces/cython/cantera/func1.pyx

```diff
@@ -46,34 +47,46 @@ cdef class Func1:
         >>> f3(6)
         30.0
 
-    For simplicity, constant functions can be defined by passing the constant
+    For simplicity, constant functions can be defined by passing a constant
     value directly::
 
         >>> f4 = Func1(2.5)
         >>> f4(0.1)
         2.5
 
     Note that all methods which accept `Func1` objects will also accept the
-    callable object and create the wrapper on their own, so it is generally
-    unnecessary to explicitly create a `Func1` object.
+    callable object and create the wrapper on their own, so it is often not
+    necessary to explicitly create a `Func1` object.
     """
-    def __cinit__(self, c):
+    def __cinit__(self, *args, **kwargs):
         self.exception = None
+        self.callable = None
+
+    def __init__(self, c):
         if hasattr(c, '__call__'):
-            self.callable = c
+            # callback function
+            self._set_callback(c)
         else:
+            arr = np.array(c)
             try:
-                # calling float() converts numpy arrays of size 1 to scalars
-                k = float(c)
-            except TypeError:
-                if hasattr(c, '__len__') and len(c) == 1:
-                    # Handle lists or tuples with a single element
+                if arr.ndim == 0:
+                    # handle constants or unsized numpy arrays
+                    k = float(c)
+                    self._set_callback(lambda t: k)
+                elif arr.size == 1:
+                    # handle lists, tuples or numpy arrays with a single element
                     k = float(c[0])
+                    self._set_callback(lambda t: k)
                 else:
-                    raise TypeError('Func1 must be constructed from a number or'
-                                    ' a callable object')
-            self.callable = lambda t: k
+                    raise TypeError
+
+            except TypeError:
+                raise TypeError(
+                    "Func1 must be constructed from a callable object "
+                    "or a number")
```

Comment:
I think here you'll want to add `raise TypeError(...) from None` to avoid the double traceback. See here: https://docs.python.org/3/library/exceptions.html